modules/katamari/src/main/ts/ephox/katamari/api/Result.ts

```diff
@@ -1,156 +1,288 @@
-    return f();
-  };
+  /**
+   * Get the value out of the inside of this `Result` object, using a default
+   * `replacement` value if the provided `Result` does not contain a value.
+   *
+   * Unlike `getOr`, in this method the replacement value is "thunked" - that is
+   * to say that you don't pass a value to `getOrThunk`, you pass a function
+   * which (if called) will **return** the value you want to use.
+   */
+  readonly getOrThunk: <T2 = T>(thunk: () => T2) => T | T2;
 
-  const map = <U>(_f: (value: T) => U) => {
-    return error<U, E>(message);
-  };
+  /**
+   * Get the value out of the inside of this `Result` object, using a default
+   * `replacement` value if the provided `Result` does not contain a value.
+   *
+   * Unlike `or`, in this method the replacement value is "thunked" - that is to
+   * say that you don't pass a value to `orThunk`, you pass a function which (if
+   * called) will **return** the value you want to use.
+   *
+   * Unlike `getOrThunk`, in this method the replacement value is also a
+   * `Result`, meaning that this method will always return a `Result`.
+   */
+  readonly orThunk: <T2 = T, E2 = E>(thunk: () => Result<T2, E2>) => Result<T | T2, E | E2>;
 
-  const mapError = <U> (f: (error: E) => U) => {
-    return error(f(message));
-  };
+  /**
+   * Get the value out of the inside of the `Result` object, throwing an
+   * exception if the provided `Result` object contains an error instead of a
+   * value.
+   *
+   * WARNING: You should only be using this function if you know that the
+   * `Result` object **does** contain a value (otherwise you're throwing
+   * exceptions in production code, which is bad).
+   *
+   * In tests this is more acceptable.
+   *
+   * Prefer other methods to this, such as `.each`.
+   */
+  readonly getOrDie: () => T;
 
-  const bind = <U>(_f: (value: T) => Result<U, E>) => {
-    return error<U, E>(message);
-  };
+  // --- Utilities ---
+
+  /**
+   * If this `Result` contains a value, perform an action on that value. Unlike
+   * the rest of the methods on this type, `.each` has side-effects. If you want
+   * to transform a `Result<T, E>` **into** something, then this is not the
+   * method for you. If you want to use a `Result<T, E>` to **do** something,
+   * then this is the method for you - provided you're okay with not doing
+   * anything in the case where the `Result` doesn't have a value inside it. If
+   * you're not sure whether your use-case fits into transforming **into**
+   * something or **doing** something, check whether it has a return value. If
+   * it does, you should be performing a transform.
+   */
+  readonly each: (worker: (value: T) => void) => void;
+
+  /**
+   * Convert this `Result<T, E>` to an `Optional<T>`. If this object contains a
+   * value, so will the output. If this object contains an error, the error will
+   * be discarded and the output will not contain a value.
+   */
+  readonly toOptional: () => Optional<T>;
+}
+
+/**
```

Comment:
```suggestion
/*
```
This doesn't appear to be a doc string and is just a general comment, so shouldn't have 2 asterisks.